File: python/RappCloud/CloudMsgs/WeatherReportCurrent.py
from RappCloud.Objects import (
    File,
    Payload)

from Cloud import (
    CloudMsg,
    CloudRequest,
    CloudResponse)


class WeatherReportCurrent(CloudMsg):
    """ Weather Report Current Cloud Message object """

    class Request(CloudRequest):
        """ Weather Report Current Cloud Request object.

        WeatherReportCurrent.Request
        """
        def __init__(self, **kwargs):
            """!
            Constructor

            @param **kwargs - Keyword arguments. Apply values to the request attributes.
                - @ref city
                - @ref weather_reporter
                - @ref metric
            """

            ## The desired city
            self.city = ''
            ## The weather API to use. Defaults to "yweather" .
            self.weather_reporter = ''
            ## The return value units.
            self.metric = 0
            super(WeatherReportCurrent.Request, self).__init__(**kwargs)


        def make_payload(self):
            """ Create and return the Payload of the Request. """
            return Payload(
                city=self.city,
                weather_reporter=self.weather_reporter,
                metric=self.metric
                )


        def make_files(self):
            """ Create and return Array of File objects of the Request. """
            return []


    class Response(CloudResponse):
        """ Weather Report Current Cloud Response object.

        WeatherReportCurrent.Response
        """
        def __init__(self, **kwargs):
            """!
            Constructor

            @param **kwargs - Keyword arguments. Apply values to the request attributes.
                - @ref error
                - @ref date
                - @ref temperature
                - @ref weather_description
                - @ref humidity
                - @ref visibility
                - @ref pressure
                - @ref wind_speed
                - @ref wind_temperature
                - @ref wind_direction
            """

            ## Error message
            self.error = ''
            ## Current date
            self.date = ''
            ## Current temperature
            self.temperature = ''
            ## A brief description of the current weather
            self.weather_description = ''
            ## Current humidity
            self.humidity = ''
            ## Current vilibility
            self.visibility = ''
            ## Current pressure
            self.pressure = ''
            ## Current speed of the wind
            self.wind_speed = ''
            ## Current temperature of the wind
            self.wind_temperature = ''
            ## Current wind direction
            self.wind_direction = ''
            super(WeatherReportCurrent.Response, self).__init__(**kwargs)


    def __init__(self, **kwargs):
        """!
        Constructor

        @param **kwargs - Keyword argumen.ts. Apply values to the request attributes.
            - @ref Request.city
            - @ref Request.weather_reporter
            - @ref Request.metric
        """

        # Create and hold the Request object for this CloudMsg
        self.req = WeatherReportCurrent.Request()
        # Create and hold the Response object for this CloudMsg
        self.resp = WeatherReportCurrent.Response()
        super(WeatherReportCurrent, self).__init__(
            svcname='weather_report_current', **kwargs)

File: js/lib/cloud/weather_report_current.js
#!/usr/bin/env node

var path = require('path');
var __cloudDir = path.join(__dirname);
var RAPPCloud = require(path.join(__cloudDir, 'RAPPCloud.js'));

/**
 * @fileOverview Prototype the RAPPCloud Service Method.
 * 
 * @class weather_report_current
 * @memberof RAPPCloud
 * @description Asynchronous Service which will request the 
 * weather_report_current RAPP platform service
 * @version 0.7.5
 * @author Lazaros Penteridis <lp@ortelio.co.uk>
 * @param city (String): The desired city
 * @param weather_reporter (String): The weather API to use. Defaults to "yweather" .
 * @param metric (Integer): The return value units.
 * @param callback is the function that will receive the result
 */
RAPPCloud.prototype.weather_report_current = function(
                                                       city, 
                                                       weather_reporter,
                                                       metric, 
                                                       callback
                                                     )
{
    var cloud = this;
    var _delegate = callback;
    var request = cloud.determine_protocol();
    var body_obj = {};
    body_obj.city = cloud.escape_string(city);
    body_obj.weather_reporter = cloud.escape_string(weather_reporter);
    body_obj.metric = metric;
    var body_json = JSON.stringify(body_obj);
    
    request.post({
                     headers: {
                         'Accept-Token' : cloud.token,
                         'Content-Type' : 'application/x-www-form-urlencoded',
                         'Connection' : 'close'
                         },
                     url: cloud.cloud_url + '/hop/weather_report_current/ ',
                     body: "json=" + body_json
                 },
                 function (error, response, json) 
                 {
                     if (!error && response.statusCode == 200)
                         handle_reply(json);
                     else if (error)
                         error_handler (error);
                     else if (response.statusCode != 200)
                         console.log ('Error: ' + response.statusCode);
                 });
    
    function handle_reply(json)
    {
        var json_obj;
        try {
			json_obj = JSON.parse(json);
			// JSON reply is: { date: '', temperature: '', weather_description: '', 
            // humidity: '', visibility: '', pressure: '', wind_speed: '', 
            // wind_temperature: '', wind_direction: '', error: ''}
			if(json_obj.error){  // Check for Errors  
				console.log('weather_report_current JSON error: ' + json_obj.error);
			}
            else {
                var weather_current = {};
                for (var prop in json_obj){
                    if (prop !== 'error')
                        weather_current[prop] = json_obj.prop;
                }
                _delegate(weather_current);
            }
        } catch (e) {
			console.log('weather_report_current::handle_reply Error parsing: ');
			return console.error(e);
		}
	}
	
	function error_handler(error) {
		return console.error(error);
	}   
};

/// Export
module.exports = RAPPCloud.weather_report_current;
